fdbserver/ClusterController.actor.cpp

```diff
@@ -2779,6 +2814,162 @@ class ClusterControllerData {
         }
     }
 
+	// Checks that if any worker or their degraded peers have recovered. If so, remove them from `workerHealth`.
+	void updateRecoveredWorkers() {
+		double currentTime = now();
+		for (auto& [workerAddress, health] : workerHealth) {
+			for (auto it = health.degradedPeers.begin(); it != health.degradedPeers.end();) {
+				if (currentTime - it->second.lastRefreshTime > SERVER_KNOBS->CC_DEGRADED_LINK_EXPIRATION_INTERVAL) {
+					TraceEvent("WorkerPeerHealthRecovered").detail("Worker", workerAddress).detail("peer", it->first);
+					health.degradedPeers.erase(it++);
+				} else {
+					++it;
+				}
+			}
+		}
+
+		for (auto it = workerHealth.begin(); it != workerHealth.end();) {
+			if (it->second.degradedPeers.empty()) {
+				TraceEvent("WorkerAllPeerHealthRecovered").detail("worker", it->first);
+				workerHealth.erase(it++);
+			} else {
+				++it;
+			}
+		}
+	}
+
+	// Returns a list of servers who are experiencing degraded links. These are candidates to perform exclusion. Note
+	// that only one endpoint of a bad link will be included in this list.
+	std::unordered_set<NetworkAddress> getServersWithDegradedLink() {
+		updateRecoveredWorkers();
+
+		// Build a map keyed by measured degraded peer. This map gives the info that who complains a particular server.
+		std::unordered_map<NetworkAddress, std::unordered_set<NetworkAddress>> degradedLinkDst2Src;
+		double currentTime = now();
+		for (const auto& [server, health] : workerHealth) {
+			for (const auto& [degradedPeer, times] : health.degradedPeers) {
+				if (currentTime - times.startTime < SERVER_KNOBS->CC_MIN_DEGRADATION_INTERVAL) {
+					// This degraded link is not long enough to be considered as degraded.
+					continue;
+				}
+				degradedLinkDst2Src[degradedPeer].insert(server);
+			}
+		}
+
+		// Sort degraded peers based on the number of workers complaining about it.
+		std::vector<std::pair<int, NetworkAddress>> count2DegradedPeer;
+		for (const auto& [degradedPeer, complainers] : degradedLinkDst2Src) {
+			count2DegradedPeer.push_back({ complainers.size(), degradedPeer });
+		}
+		std::sort(count2DegradedPeer.begin(), count2DegradedPeer.end(), std::greater<>());
```

Comment:
I don't know what `std::greater<>()` does here. Does it do some magic sorting the pairs on the first element (i.e., int)?

Reply:
This sorts the vector in decreasing order.

Reply:
Yeah but the vector's elements are pairs, why can `std::greater<>()` work without a custom comparator defined?

Reply:
pairs has comparator (<, >, etc.) built in.

All `greater` does is to simply return p1 > p2 [[ref](https://www.cplusplus.com/reference/functional/greater/)]